cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -229,6 +604,710 @@ std::unique_ptr<KernelState> AllInit(KernelContext*, const KernelInitArgs& args)
+      total_length += offsets_batch_[i];
+      offsets_batch_[i] = total_length_before;
+    }
+    offsets_batch_[batch.length] = total_length;
+
+    key_bytes_batch_.clear();
+    key_bytes_batch_.resize(total_length);
+    key_buf_ptrs_.clear();
+    key_buf_ptrs_.resize(batch.length);
+    for (int64_t i = 0; i < batch.length; ++i) {
+      key_buf_ptrs_[i] = key_bytes_batch_.data() + offsets_batch_[i];
+    }
+    for (size_t i = 0; i < keys.size(); ++i) {
+      encode_next_impl[i].encode_next_impl(keys[i], key_buf_ptrs_.data());
+    }
+
+    group_ids_batch_.clear();
+    group_ids_batch_.resize(batch.length);
+    for (int64_t i = 0; i < batch.length; ++i) {
+      int32_t key_length = offsets_batch_[i + 1] - offsets_batch_[i];
+      std::string key(
+          reinterpret_cast<const char*>(key_bytes_batch_.data() + offsets_batch_[i]),
+          key_length);
+      auto iter = map_.find(key);
+      if (iter == map_.end()) {
+        group_ids_batch_[i] = n_groups++;
+        auto next_key_offset = static_cast<int32_t>(key_bytes_.size());
+        key_bytes_.resize(next_key_offset + key_length);
+        offsets_.push_back(next_key_offset + key_length);
+        memcpy(key_bytes_.data() + next_key_offset, key.c_str(), key_length);
+        map_.insert(std::make_pair(key, group_ids_batch_[i]));
+      } else {
+        group_ids_batch_[i] = iter->second;
+      }
+    }
+
+    for (size_t i = 0; i < aggregators.size(); ++i) {
+      aggregators[i]->Consume(ctx, aggregands[i], group_ids_batch_.data());
+      if (ctx->HasError()) return;
+    }
+  }
+
+  void MergeFrom(KernelContext* ctx, KernelState&& src) override {
+    // TODO(ARROW-11840) merge two hash tables
+    ctx->SetStatus(Status::NotImplemented("merging grouped aggregations"));
+  }
+
+  void Finalize(KernelContext* ctx, Datum* out) override {
+    size_t n_keys = decode_next_impl.size();
+    ArrayDataVector out_columns(aggregators.size() + n_keys);
+    for (size_t i = 0; i < aggregators.size(); ++i) {
+      Datum aggregand;
+      aggregators[i]->Finalize(ctx, &aggregand);
+      if (ctx->HasError()) return;
+      out_columns[i] = aggregand.array();
+    }
+
+    key_buf_ptrs_.clear();
+    key_buf_ptrs_.resize(n_groups);
+    for (int64_t i = 0; i < n_groups; ++i) {
+      key_buf_ptrs_[i] = key_bytes_.data() + offsets_[i];
+    }
+
+    int64_t length = n_groups;
+    for (size_t i = 0; i < n_keys; ++i) {
+      std::shared_ptr<ArrayData> key_array;
+      decode_next_impl[i].decode_next_impl(ctx, static_cast<int32_t>(length),
+                                           key_buf_ptrs_.data(), &key_array);
+      out_columns[aggregators.size() + i] = std::move(key_array);
+    }
+
+    *out = ArrayData::Make(std::move(out_type), length, {/*null_bitmap=*/nullptr},
+                           std::move(out_columns));
+  }
+  std::vector<int32_t> offsets_batch_;
+  std::vector<uint8_t> key_bytes_batch_;
+  std::vector<uint8_t*> key_buf_ptrs_;
```

Comment:
As an optimization, we may also want a `std::vector<int64_t> key_null_counts_`... though this may not be beneficial.

Reply:
IIUC, this would be supported by using the "count" aggregation